In [1]:
import requests
import pandas as pd

# Define the URL
url = "https://data.cityofnewyork.us/resource/43nn-pn8j.json"

# Initialize an empty list to store data
data_list = []

# Initialize offset
offset = 0

# Define limit
limit = 1000

# Make API calls until all data is fetched
while True:
    # Construct the URL with the offset
    params = {"$limit": limit, "$offset": offset}
    response = requests.get(url, params=params)
    
    # Check if the response is successful
    if response.status_code == 200:
        # Append the fetched data to the list
        data_list.extend(response.json())
        
        # Increment the offset
        offset += limit
        
        # If the length of fetched data is less than the limit, it means we have reached the end of the dataset
        if len(response.json()) < limit:
            break
    else:
        print("Error occurred while fetching data")
        break

# Convert the list of dictionaries to a DataFrame
restos = pd.DataFrame(data_list)

# Display the first few rows of the DataFrame
print(restos.head())

      camis           boro building               street zipcode       phone  \
0  50147001  Staten Island      700     ARTHUR KILL ROAD   10308  6466109071   
1  50148665      Manhattan     1415             2 AVENUE   10021  9179233223   
2  50124392       Brooklyn     1018             AVENUE M   11230  9174149203   
3  50141395      Manhattan      827             BROADWAY   10003  3022986500   
4  50133652       Brooklyn      752  CONEY ISLAND AVENUE   11218  6463024102   

           inspection_date   critical_flag              record_date  \
0  1900-01-01T00:00:00.000  Not Applicable  2024-03-21T06:00:13.000   
1  1900-01-01T00:00:00.000  Not Applicable  2024-03-21T06:00:13.000   
2  1900-01-01T00:00:00.000  Not Applicable  2024-03-21T06:00:13.000   
3  1900-01-01T00:00:00.000  Not Applicable  2024-03-21T06:00:13.000   
4  1900-01-01T00:00:00.000  Not Applicable  2024-03-21T06:00:13.000   

          latitude  ...   nta                   dba cuisine_description  \
0  40.56018384638

In [2]:
# Get the shape of the df
num_rows = restos.shape[0]

# Print the number of rows
print("Number of rows:", num_rows)

# Get the column names and data types
column_info = restos.dtypes

# Print the column names and data types
print("Column names and data types:")
for column_name, dtype in column_info.items():
    print(f"{column_name}: {dtype}")

Number of rows: 220598
Column names and data types:
camis: object
boro: object
building: object
street: object
zipcode: object
phone: object
inspection_date: object
critical_flag: object
record_date: object
latitude: object
longitude: object
community_board: object
council_district: object
census_tract: object
bin: object
bbl: object
nta: object
dba: object
cuisine_description: object
action: object
violation_code: object
violation_description: object
score: object
inspection_type: object
grade: object
grade_date: object


In [3]:
# Count the number of N/A values in each column
na_counts = restos.isna().sum()

# Print the column names and their corresponding N/A counts
print("Column names and N/A counts:")
for column_name, na_count in na_counts.items():
    print(f"{column_name}: {na_count}")

Column names and N/A counts:
camis: 0
boro: 0
building: 425
street: 0
zipcode: 2746
phone: 2
inspection_date: 0
critical_flag: 0
record_date: 0
latitude: 287
longitude: 287
community_board: 3416
council_district: 3411
census_tract: 3411
bin: 4462
bbl: 670
nta: 3416
dba: 552
cuisine_description: 2311
action: 2311
violation_code: 3451
violation_description: 3451
score: 10580
inspection_type: 2311
grade: 112086
grade_date: 120900


In [4]:
# Display the top 5 rows of the 'inspection_date' column
top_5_inspection_dates = restos['inspection_date'].head(5)
print(top_5_inspection_dates)

0    1900-01-01T00:00:00.000
1    1900-01-01T00:00:00.000
2    1900-01-01T00:00:00.000
3    1900-01-01T00:00:00.000
4    1900-01-01T00:00:00.000
Name: inspection_date, dtype: object


In [5]:
# Display the bottom 10 rows of the 'inspection_date' column
bottom_10_inspection_dates = restos['inspection_date'].tail(10)
print(bottom_10_inspection_dates)

220588    2023-02-07T00:00:00.000
220589    2024-01-04T00:00:00.000
220590    2022-11-07T00:00:00.000
220591    2022-08-17T00:00:00.000
220592    2023-03-08T00:00:00.000
220593    2023-03-29T00:00:00.000
220594    2023-04-12T00:00:00.000
220595    2022-04-22T00:00:00.000
220596    2022-05-26T00:00:00.000
220597    2023-06-02T00:00:00.000
Name: inspection_date, dtype: object


In [8]:
# Convert 'inspection_date' column to datetime format
restos['inspection_date'] = pd.to_datetime(restos['inspection_date'])

# Count the rows where 'inspection_date' equals '1900-01-01'
count = (restos['inspection_date'] == '1900-01-01').sum()

print("Number of rows with inspection_date equals '1900-01-01':", count)

Number of rows with inspection_date equals '1900-01-01': 2311


In [9]:
# Create a new DataFrame 'restos_2' excluding rows with inspection_date equals '1900-01-01'
restos_2 = restos[restos['inspection_date'] != '1900-01-01']

# Reset index of the new DataFrame
restos_2.reset_index(drop=True, inplace=True)

# Now, 'restos_2' contains rows excluding '1900-01-01' in the 'inspection_date' column
print(restos_2)

           camis       boro building                     street zipcode  \
0       50112427   Brooklyn    1801C                   AVENUE U     NaN   
1       41108575     Queens      NaN  JFK INTERNATIONAL AIRPORT   11430   
2       41194745     Queens    14009             CHERRRY AVENUE     NaN   
3       41641108     Queens    11410               SUTPHIN BLVD   11434   
4       50084728  Manhattan      787                   7 AVENUE   10019   
...          ...        ...      ...                        ...     ...   
218282  50080925   Brooklyn      668                   3 AVENUE   11232   
218283  50105008  Manhattan      200                   BROADWAY   10038   
218284  41462542   Brooklyn      788               UNION STREET   11215   
218285  50033000   Brooklyn     7206                   3 AVENUE   11209   
218286  41471348   Brooklyn      263          NORTH    6 STREET   11211   

             phone inspection_date   critical_flag              record_date  \
0       6462678950  

In [10]:
num_rows_restos_2 = restos_2.shape[0]
print("Number of rows in restos_2:", num_rows_restos_2)

Number of rows in restos_2: 218287


In [11]:
# Use unique() function to get unique values in the 'inspection_date' column
unique_inspection_dates = restos_2['inspection_date'].unique()

print("Unique inspection dates:")
print(unique_inspection_dates)

Unique inspection dates:
['2023-08-01T00:00:00.000000000' '2018-11-14T00:00:00.000000000'
 '2022-02-15T00:00:00.000000000' ... '2015-10-15T00:00:00.000000000'
 '2017-05-31T00:00:00.000000000' '2021-06-29T00:00:00.000000000']


In [12]:
# Find the minimum and maximum dates
min_date = restos_2['inspection_date'].min()
max_date = restos_2['inspection_date'].max()

print("Range of dates in the 'inspection_date' column:")
print("Minimum date:", min_date)
print("Maximum date:", max_date)

Range of dates in the 'inspection_date' column:
Minimum date: 2015-09-24 00:00:00
Maximum date: 2024-03-19 00:00:00


In [13]:
# Get the shape of the df
num_rows_2 = restos_2.shape[0]

# Print the number of rows
print("Number of rows:", num_rows_2)
# Number of rows: 217475

# Get the column names and data types
column_info_2 = restos_2.dtypes

# Print the column names and data types
print("Column names and data types:")
for column_name, dtype in column_info_2.items():
    print(f"{column_name}: {dtype}")

Number of rows: 218287
Column names and data types:
camis: object
boro: object
building: object
street: object
zipcode: object
phone: object
inspection_date: datetime64[ns]
critical_flag: object
record_date: object
latitude: object
longitude: object
community_board: object
council_district: object
census_tract: object
bin: object
bbl: object
nta: object
dba: object
cuisine_description: object
action: object
violation_code: object
violation_description: object
score: object
inspection_type: object
grade: object
grade_date: object


So we started with 220,598 rows in the original dataframe "restos". And after removing all the rows that have a value of "1900-01-01" in the "inspection_date" column, we are left with 218,287 rows. This means 2311 restaurants in the dataset "restos" are new establishments that have not yet received an inspection. 

In [19]:
# Display the top 10 rows of the 'action' column
top_10_action = restos_2['action'].head(10)
print(top_10_action)

0      Violations were cited in the following area(s).
1      Violations were cited in the following area(s).
2      Violations were cited in the following area(s).
3    No violations were recorded at the time of thi...
4      Violations were cited in the following area(s).
5    No violations were recorded at the time of thi...
6      Violations were cited in the following area(s).
7      Violations were cited in the following area(s).
8                    Establishment re-opened by DOHMH.
9      Violations were cited in the following area(s).
Name: action, dtype: object


In [20]:
# Display the bottom 10 rows of the 'action' column
bottom_10_action = restos_2['action'].tail(10)
print(bottom_10_action)

218277      Violations were cited in the following area(s).
218278      Violations were cited in the following area(s).
218279      Violations were cited in the following area(s).
218280      Violations were cited in the following area(s).
218281    Establishment Closed by DOHMH. Violations were...
218282      Violations were cited in the following area(s).
218283    Establishment Closed by DOHMH. Violations were...
218284      Violations were cited in the following area(s).
218285      Violations were cited in the following area(s).
218286      Violations were cited in the following area(s).
Name: action, dtype: object


In [27]:
# Filter out rows where the 'violation_code' column contains NaN or NA values
filtered_rows_violation_code = restos_2[restos_2['violation_code'].isna()]

# Calculate the number of rows in the filtered dataframe
num_filtered_rows = filtered_rows_violation_code.shape[0]
print("Number of filtered rows:", num_filtered_rows)

# Get the unique values in the 'action' column from the filtered rows
unique_actions = filtered_rows_violation_code['action'].unique()

# Display the unique values in the 'action' column
print(unique_actions)

Number of filtered rows: 1140
['No violations were recorded at the time of this inspection.'
 'Establishment re-opened by DOHMH.'
 'Violations were cited in the following area(s).'
 'Establishment Closed by DOHMH. Violations were cited in the following area(s) and those requiring immediate action were addressed.']


In [28]:
# Remove rows with NaN or NA values in the "violation_code" column
restos_2_cleaned = restos_2.dropna(subset=['violation_code'])

# Reset index of the new DataFrame
restos_2_cleaned.reset_index(drop=True, inplace=True)

# Now, 'restos_2_cleaned' contains rows without NaN or NA values in the "violation_code" column
print(restos_2_cleaned)

           camis       boro building                     street zipcode  \
0       50112427   Brooklyn    1801C                   AVENUE U     NaN   
1       41108575     Queens      NaN  JFK INTERNATIONAL AIRPORT   11430   
2       41194745     Queens    14009             CHERRRY AVENUE     NaN   
3       50084728  Manhattan      787                   7 AVENUE   10019   
4       50012812     Queens    90-15           QUEENS BOULEVARD   11373   
...          ...        ...      ...                        ...     ...   
217142  50080925   Brooklyn      668                   3 AVENUE   11232   
217143  50105008  Manhattan      200                   BROADWAY   10038   
217144  41462542   Brooklyn      788               UNION STREET   11215   
217145  50033000   Brooklyn     7206                   3 AVENUE   11209   
217146  41471348   Brooklyn      263          NORTH    6 STREET   11211   

             phone inspection_date critical_flag              record_date  \
0       6462678950    

In [29]:
# Get the number of rows in the DataFrame
num_rows = restos_2_cleaned.shape[0]

print("Number of rows in 'restos_2_cleaned':", num_rows)

Number of rows in 'restos_2_cleaned': 217147


In [31]:
# Count the number of NaN or NA values in the "violation_code" column
num_nan_values = restos_2_cleaned['violation_code'].isna().sum()

print("Number of NaN or NA values in the 'violation_code' column of 'restos_2_cleaned':", num_nan_values)

Number of NaN or NA values in the 'violation_code' column of 'restos_2_cleaned': 0


In [35]:
# Get the unique values in the "violation_code" column
unique_violation_codes = restos_2_cleaned['violation_code'].unique()

# Count the number of unique values
num_unique_violation_codes = len(unique_violation_codes)

print("Number of unique values in the 'violation_code' column:", num_unique_violation_codes)

print("Unique values in the 'violation_code' column:")
print(unique_violation_codes)

print("Unique values in the 'violation_code' column:")
for code in unique_violation_codes:
    print(code)

Number of unique values in the 'violation_code' column: 143
Unique values in the 'violation_code' column:
['04M' '09B' '09C' '02B' '10E' '10J' '04L' '22F' '20-06' '20F' '10H' '06F'
 '06E' '15F6' '09E' '10C' '04A' '04K' '02A' '10B' '02G' '08A' '10F' '05H'
 '10D' '03A' '04H' '16-03' '02H' '05D' '06C' '20-01' '06D' '06A' '04N'
 '04D' '19-04' '20-08' '16-04' '15-21' '20-04' '05A' '04C' '19-07' '10G'
 '05F' '08C' '06B' '08B' '19-06' '15-27' '04J' '03B' '03I' '02C' '16-02'
 '22A' '28-03' '10A' '15-37' '28-01' '18-11' '28-06' '16B' '10I' '05E'
 '18G' '05C' '19-10' '20A' '04E' '02I' '06G' '20D' '18F' '16-06' '15E2'
 '28-05' '04O' '22E' '28-07' '04P' '09A' '15F2' '19-08' '22G' '22C' '16E'
 '05B' '04F' '28-04' '18-13' '15-22' '03F' '19-05' '16D' '15F7' '20E'
 '06I' '03E' '15-33' '19-11' '20-05' '16-01' '09D' '15-42' '18C' '18-12'
 '15-01' '18D' '04B' '15L' '16C' '02F' '18-14' '16A' '02D' '18-01' '06H'
 '15-39' '03D' '07A' '15-36' '03C' '15-17' '15A1' '20-07' '18B' '20C'
 '15E3' '15F1' '03G' '18-

In [36]:
restos_3 = restos_2_cleaned

In [37]:
# Get the shape of the df
num_rows = restos_3.shape[0]

# Print the number of rows
print("Number of rows:", num_rows)

# Get the column names and data types
column_info = restos_3.dtypes

# Print the column names and data types
print("Column names and data types:")
for column_name, dtype in column_info.items():
    print(f"{column_name}: {dtype}")

Number of rows: 217147
Column names and data types:
camis: object
boro: object
building: object
street: object
zipcode: object
phone: object
inspection_date: datetime64[ns]
critical_flag: object
record_date: object
latitude: object
longitude: object
community_board: object
council_district: object
census_tract: object
bin: object
bbl: object
nta: object
dba: object
cuisine_description: object
action: object
violation_code: object
violation_description: object
score: object
inspection_type: object
grade: object
grade_date: object


In [39]:
# Count the number of N/A values in each column
na_counts = restos_3.isna().sum()

# Print the column names and their corresponding N/A counts
print("Column names and N/A counts:")
for column_name, na_count in na_counts.items():
    print(f"{column_name}: {na_count}")

Column names and N/A counts:
camis: 0
boro: 0
building: 399
street: 0
zipcode: 2648
phone: 0
inspection_date: 0
critical_flag: 0
record_date: 0
latitude: 248
longitude: 248
community_board: 3266
council_district: 3261
census_tract: 3261
bin: 4288
bbl: 618
nta: 3266
dba: 0
cuisine_description: 0
action: 0
violation_code: 0
violation_description: 0
score: 7753
inspection_type: 0
grade: 109074
grade_date: 117862


In [40]:
# Remove rows with NaN or NA values in both "latitude" and "longitude" columns
restos_4 = restos_3.dropna(subset=['latitude', 'longitude'], how='all')

# Reset index of the new DataFrame
restos_4.reset_index(drop=True, inplace=True)

# Now, 'restos_3_cleaned' contains rows without NaN or NA values in both "latitude" and "longitude" columns
print(restos_4)

           camis       boro building                     street zipcode  \
0       50112427   Brooklyn    1801C                   AVENUE U     NaN   
1       41108575     Queens      NaN  JFK INTERNATIONAL AIRPORT   11430   
2       41194745     Queens    14009             CHERRRY AVENUE     NaN   
3       50084728  Manhattan      787                   7 AVENUE   10019   
4       50012812     Queens    90-15           QUEENS BOULEVARD   11373   
...          ...        ...      ...                        ...     ...   
216894  50080925   Brooklyn      668                   3 AVENUE   11232   
216895  50105008  Manhattan      200                   BROADWAY   10038   
216896  41462542   Brooklyn      788               UNION STREET   11215   
216897  50033000   Brooklyn     7206                   3 AVENUE   11209   
216898  41471348   Brooklyn      263          NORTH    6 STREET   11211   

             phone inspection_date critical_flag              record_date  \
0       6462678950    

In [41]:
# Get the unique values in the 'boro' column
unique_boro_values = restos_4['boro'].unique()

# Print the unique values
print("Unique values in the 'boro' column:", unique_boro_values)

Unique values in the 'boro' column: ['Brooklyn' 'Queens' 'Manhattan' 'Staten Island' 'Bronx']


In [43]:
# Get the unique values in the "dba" column
unique_dbas = restos_4['dba'].unique()

# Count the number of unique values
num_unique_dbas = len(unique_dbas)

print("Number of unique values in the 'dba' column:", num_unique_dbas)

print("Unique values in the 'dba' column:")
for dba in unique_dbas:
    print(dba)

Number of unique values in the 'dba' column: 20913
Unique values in the 'dba' column:
HAPPY CANTEEN
SOHO BISTRO
GOLDEN PALACE GOURMET
SIDLEY AUSTIN LLP
PANDA EXPRESS
AER LINGUS LOUNGE
TOAST & ROAST
MCDONALD'S
BAR SIX
BIRDS OF A FEATHER
CHOCK FULL O' NUT
CUCHIFRITO
THE SICILIAN
DANGOL
POPEYES
SWEETGREEN
LESLIE'S KITCHEN
VIN SUR VINGT
SAN REMO PIZZERIA
BLUE MAIZ
PORCELAIN
T & G FLAVORS
PARIS BAGUETTE
NEW HO WAH CHINESE
PAPA'S PIZZA
98K HAMBURGER
NATTO HIBACHI & SUSHI
TRINI BREAKFAST SHED II
PIZZA D'AMORE
NEW WIN HING III CHINESE RESTAURANT
BABA COOL
LOS GIRASOLES BAKERY
SOCIAL HOUSE
M & O BAGELS
BRAVAZO
MUNDO FELIZ
TAKUMI OMAKASE
BAO & PANCAKE
SUSHI OF GARI TRIBECA
PANSHI
TEN THOUSAND COFFEE
FIVE GUYS
TANDOORI FOOD & BAKERY
GYU-KAKU
CHONGQING WHARF
THE LITTLE BEET
ATOMIC WINGS
SIMPLE LOAF BAKEHOUSE
FOO-HING KITCHEN
BRANDY LIBRARY LOUNGE
PAUSE CAFE
BAR VELOCE
HO'BRAH
TRATTORIA BIANCA
DUNKIN
CAFE MOGADOR
CAP'T LOUI
KORN'S BAKERY
FINE & RARE
BIRRIA LES TACOS
MEE SUM CAFE
LUPITA'S MEXICAN FO

In [45]:
# Filter rows where both "latitude" and "longitude" are 0
zero_coords_rows = restos_4[(restos_4['latitude'] == 0) & (restos_4['longitude'] == 0)]

print(zero_coords_rows)

Empty DataFrame
Columns: [camis, boro, building, street, zipcode, phone, inspection_date, critical_flag, record_date, latitude, longitude, community_board, council_district, census_tract, bin, bbl, nta, dba, cuisine_description, action, violation_code, violation_description, score, inspection_type, grade, grade_date]
Index: []

[0 rows x 26 columns]


In [44]:
# Write the dataframe "restos_4" to a CSV file named "restos_4.csv"
restos_4.to_csv("restos_4.csv", index=False)